### PR TITLE
chore: added bencher for relative benchmarking

### LINKED ISSUE #1300: Integrate bencher for benchmarks
The idea here is to integrate [Bencher] by @epompeii and keep track of performance.

[Bencher]: https://github.com/bencherdev/bencher

It would be great if we could push data of older commits for the last 2-3 months and visualize it on something like — https://bencher.dev/perf/advent-of-code

### PR DESCRIPTION
**Pull request by alankritdabral**
**Summary:**  
Added bencher for tracking benchmarks and relative benchmarking 

As using bencher.dev with `crago criterion` was showing error refer [comment](https://github.com/tailcallhq/tailcall/issues/1300#issuecomment-2005526402)
But  bencher.dev is more reliable in long term use so i have created this pr.

Results after merging this pr would look like  [results](https://github.com/alankritdabral/tailcall/actions/runs/8627520412/job/23648062964)
New relative benchmarking  [results](https://github.com/alankritdabral/tailcall/pull/114)

_If a maintainer wants to test this he can simply create a test pr with changed benchmarks.yml just like [benchmarks.txt](https://github.com/tailcallhq/tailcall/files/14929139/benchmarks.txt) I cannot try this because i does not have permission to use secrets_

Workflow:
[Track_and_Cache_Benchmarks](https://github.com/tailcallhq/tailcall/pull/1679/files#diff-aacbd0338bcf598920c120b69adaba2672acc7ea314cfca0f792bebe5fe2acd0R72) job;-

- Runs on push to main branch
- Runs cargo bench and save results in `base.txt`
- Save and track benchmarks using bencher.dev
- Save `base.txt` as cache to use in relative-benchmarking  
 
[Criterion_compare](https://github.com/tailcallhq/tailcall/pull/1679/files#diff-aacbd0338bcf598920c120b69adaba2672acc7ea314cfca0f792bebe5fe2acd0R111) job;-

- Restore `base.txt` cache
- Runs cargo bench and save results in `change.txt`
- Create feature-branch with Base benchmarks(`base.txt`) and reset its history on each run
- Create a threshold for the feature-branch
- Compare benchmarks feature-branch with `change.txt` and prints the result  

**Issue Reference(s):**  
Fixes  #1300 
Claim #1300  
**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [x] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

## Summary by CodeRabbit

- **Chores**
	- Updated benchmarking workflow to enhance performance tracking.
- **Documentation**
	- Revised contributing guidelines to emphasize observability and documentation enhancements.
- **New Features**
	- Renamed and optimized benchmarking jobs for improved performance monitoring.
	- Streamlined benchmarking commands with Bencher CLI integration.
	- Restructured caching paths and added environment variables for better benchmarking comparison.
	- Enhanced steps and commands for benchmarking comparison with Bencher CLI.

### CODE REVIEW

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -103,33 +116,56 @@ jobs:
...
+          --token "${{ secrets.BENCHER_API_TOKEN }}" \
+          --file "$BASE_BENCHMARK_RESULTS"
+
+      - name: Create PR threshold
+        run: |
+          bencher threshold create \
+          --branch '${{ github.head_ref }}_commit_${{ github.event.pull_request.head.sha }}' \
+          --testbed "$BENCHER_TESTBED" \
+          --measure latency \
+          --test percentage \
```
Comment (alankritdabral) [inline .github/workflows/benchmark.yml:160]: https://bencher.dev/docs/explanation/thresholds/#percentage-thresholds

File: .github/workflows/benchmark.yml [hunk 2]
```diff
@@ -103,33 +116,56 @@ jobs:
...
+
+      - name: Install Bencher CLI
+        uses: bencherdev/bencher@main
 
-      - name: Check Degradation
-        run: ./scripts/criterion_compare.rs  benches/main_benchmarks.json benches/benchmarks.json check
+      - name: Track base benchmarks
+        run: |
+          bencher run \
+          --if-branch '${{ github.head_ref }}_commit_${{ github.event.pull_request.head.sha }}' \
```
Comment (alankritdabral) [inline .github/workflows/benchmark.yml:149]: having same branch name for each threshold returns an error so i am changing branch name with each commit

File: .github/workflows/benchmark.yml [hunk 3]
```diff
@@ -103,33 +115,57 @@ jobs:
     permissions:
       pull-requests: write
       contents: write
+    env:
+      BENCHER_PROJECT: tailcall
+      BENCHER_ADAPTER: rust_criterion
+      BENCHER_TESTBED: benchmarking-runner
+      PR_BENCHMARK_RESULTS: benches/change.txt
+      BASE_BENCHMARK_RESULTS: benches/base.txt
+      BRANCH_NAME: ${{ github.head_ref }}_commit_${{ github.event.pull_request.head.sha }}
```
Comment (alankritdabral) [inline .github/workflows/benchmark.yml:124]: In Create PR threshold job 
for each run if branch name is same an error is returned saying identical branch_Id
so i am changing branch name with each commit

File: .github/workflows/benchmark.yml [hunk 4]
```diff
@@ -103,33 +115,57 @@ jobs:
...
+          --upper-boundary ${{ env.UPPER_BOUNDARY }} \
+          --token "${{ secrets.BENCHER_API_TOKEN }}"
+
+      - name: Track PR Benchmarks
+        run: |
+          bencher run \
+          --branch "$BRANCH_NAME" \
+          --token "${{ secrets.BENCHER_API_TOKEN }}" \
+          --github-actions "${{ secrets.GITHUB_TOKEN }}" \
+          --err \
```
Comment (alankritdabral) [inline .github/workflows/benchmark.yml:170]: if any benchmarks exceeds 10 % the ci fails

**Comment by alankritdabral** [comment]
@tusharmath can i get a review on this pr? :smile_cat:

File: .github/workflows/benchmark.yml [hunk 5]
```diff
@@ -5,7 +5,7 @@ on:
     paths-ignore: ["docs/**", "**.md"]
     branches:
       - main
-  pull_request:
+  pull_request_target:
```
Comment (alankritdabral) [inline .github/workflows/benchmark.yml:8]: `pull_request_target` to use `secrets.GITHUB_TOKEN` for forked repos

File: .github/workflows/benchmark.yml [hunk 6]
```diff
@@ -16,7 +16,7 @@ concurrency:
 jobs:
   macro_benchmark:
     name: Macro Benchmarks
-    if: "contains(github.event.pull_request.labels.*.name, 'ci: benchmark') || github.event_name == 'push'"
+    if: "contains(github.event.pull_request.labels.*.name, 'ci: benchmark')  || github.event_name == 'push'"
```
Comment (alankritdabral) [inline .github/workflows/benchmark.yml:19]:
```suggestion
    if: "contains(github.event.pull_request.labels.*.name, 'ci: benchmark') || github.event_name == 'push'"
```

File: .github/workflows/benchmark.yml [hunk 7]
```diff
@@ -103,33 +115,57 @@ jobs:
...
+
+      - name: Install Bencher CLI
+        uses: bencherdev/bencher@main
 
-      - name: Check Degradation
-        run: ./scripts/criterion_compare.rs  benches/main_benchmarks.json benches/benchmarks.json check
+      - name: Track base benchmarks
+        run: |
+          bencher run \
+          --if-branch "$BRANCH_NAME" \
```
Comment (epompeii) [inline .github/workflows/benchmark.yml:149]: This option is now deprecated. Just use `--branch` instead: https://bencher.dev/docs/explanation/branch-selection/#--branch-branch

File: .github/workflows/benchmark.yml [hunk 8]
```diff
@@ -69,31 +69,43 @@ jobs:
...
+      - name: Install Bencher CLI
+        uses: bencherdev/bencher@main
 
-      - name: Cache Criterion Benchmarks Json
+      - name: Run and track Benchmarks
+        run: |
+          cargo bench > benches/base.txt
+          bencher run \
+          --project "$BENCHER_PROJECT" \
+          --branch main \
```
Comment (epompeii) [inline .github/workflows/benchmark.yml:99]: Use the new `--branch-reset` option: https://bencher.dev/docs/explanation/branch-selection/#--branch-reset

See: https://bencher.dev/docs/how-to/track-benchmarks/#relative-continuous-benchmarking

File: .github/workflows/benchmark.yml (hunk 3 above)
Comment (epompeii) [inline .github/workflows/benchmark.yml:124]: Creating a new branch name yourself should no longer be necessary, see: https://bencher.dev/docs/how-to/track-benchmarks/#relative-continuous-benchmarking

File: .github/workflows/benchmark.yml [hunk 9]
```diff
@@ -5,7 +5,7 @@ on:
     paths-ignore: ["docs/**", "**.md"]
     branches:
       - main
-  pull_request:
+  pull_request_target:
     paths-ignore: ["docs/**", "**.md"]
     types: [opened, reopened, synchronize, labeled]
```
Comment (epompeii) [inline .github/workflows/benchmark.yml:10]: Do you want `edited` in case the base branch changes?

**Review by epompeii — CHANGES_REQUESTED** [review]

File: .github/workflows/benchmark.yml (hunk 8 above)
Comment (alankritdabral) [reply .github/workflows/benchmark.yml:99]: ` Run and track Benchmarks ` job is to track the benchmarks so we have a  performance history of the benchmarks.
Using --branch-reset would recreate the branch and discontinue the tracking.
I think you meant using  `--branch-reset` in the [criterion_compare](https://github.com/tailcallhq/tailcall/pull/1679/files#diff-aacbd0338bcf598920c120b69adaba2672acc7ea314cfca0f792bebe5fe2acd0L99) job

File: .github/workflows/benchmark.yml (hunk 9 above)
Comment (alankritdabral) [reply .github/workflows/benchmark.yml:10]: Unsure about this one

**Pull request closed without merging**